### PR TITLE
Adds ability to watch files and exit gracefully if they change

### LINKED ISSUE #16: Live option reload
We should be able to live reload options for kube-oidc-proxy, i.e. live reload certs when the file changes

### PR DESCRIPTION
**Pull request by JoshVanL**
```release-note
Adds options to enable watching files that cause a graceful exit when modified
```

In favor of #88 

/hold
/assign
fixes #16
/milestone v0.2

### CODE REVIEW

**Comment by JoshVanL** [comment]
Should be good to have a look now
/unassign
/assign @simonswine 
/hold cancel
/test pull-kube-oidc-proxy-e2e-v1-14
/test pull-kube-oidc-proxy-e2e-v1-13
/test pull-kube-oidc-proxy-e2e-v1-12
/test pull-kube-oidc-proxy-e2e-v1-11

File: test/e2e/suite/cases/watch/watch.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+		podUID := pod.ObjectMeta.UID
+
+		By("Update ConfigMap Data")
+		cm.Data["key-1"] = "this is different data"
+		cm.Data["key-2"] = "this is more different data"
+		_, err = f.Helper().KubeClient.CoreV1().ConfigMaps(f.Namespace.Name).Update(cm)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for proxy to pick up change")
+		time.Sleep(time.Second * 15)
```
Comment (simonswine) [inline test/e2e/suite/cases/watch/watch.go]: Not a big fan of sleep in testing. Can we use a retry mechanism?

File: test/e2e/suite/cases/watch/watch.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		podUID := pod.ObjectMeta.UID
+
+		By("Update Secret Data")
+		sec.Data["key-1"] = []byte("this is different data")
+		sec.Data["key-2"] = []byte("this is more different data")
+		_, err = f.Helper().KubeClient.CoreV1().Secrets(f.Namespace.Name).Update(sec)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for proxy to pick up change")
+		time.Sleep(time.Second * 15)
```
Comment (simonswine) [inline test/e2e/suite/cases/watch/watch.go]: Not a big fan of sleep in testing. Can we use a retry mechanism?

**Comment by JoshVanL** [comment]
@simonswine 
I have updated the docs to make it clear it will terminate the proxy on watch change and mentioned the HPA issue.
I think the CLI flag is probably clear enough the exit behavior.

`Comma-separated list of files to watch for changes. If a change is detected then send self a SIGHUP signal and exit gracefully.` 

I've updated the tests and will re-assign once they've passed.

**Comment by JoshVanL** [comment]
This functionality should be checked whether this has been included, and if not, enabled using apiserver module

/close

**Comment by jescarri** [comment]
What's the status on this PR? is this a feature that's not going to be included in future releases?

Thanks!

**Comment by JoshVanL** [comment]
Hey @jescarri,

In theory, now that we have updated dependencies on k8s to 1.18, this should be included as default. I haven't tested this however, so if it isn't the case than I will do some investigating to make sure it's there.

**Pull request closed without merging**